### ISSUE TITLE
Resource limits for RAM, dimensions, and CPU time

### ISSUE THREAD

**Issue by carbotaniuman** [shssoichiro/oxipng#411]
I use oxipng to process untrusted files from uploaded users, and we currently rely on pod level resource limits in order to protect against PNG bombs. Is it possible for there to be resource limits by oxipng itself, ala ImageMagick?

**Comment by shssoichiro** [comment]
Since Rust doesn't run within a VM (ala Java), I don't think there's a way to set explicit memory limits, but there could be options added to limit the width and height. There is an existing CLI option, `--timeout`, which will limit the wall time oxipng will spend on a file.

I'm reading that these attacks can also happen through excessively large iCCP chunks, zTXt chunks, and iTXt chunks, so ensuring oxipng protects against these (by default) would be a good addition as well.

**Comment by Winterhuman** [comment]
For anyone looking for a temporary workaround, you can use `systemd-run` to do resource limiting, e.g.

`systemd-run --user --wait -Pp MemoryHigh=5% oxipng input.png output.png`

You can set `-p MemoryMax=10%` as well if you want the transient service to exit if its memory usage creeps too high (`MemoryHigh` is only a soft cap).

**Comment by Maingron** [comment]
I had my own take on this in an issue, closed as duplicate (#722):

## Condensed version of my take:

### General Idea

It would be nice to have the option to exclude files that have a larger dimension than `--max-dimensions`.

I like to apply the most heavy settings on smaller images like icons (`oxipng  -o max -Z --preserve --alpha -f 0-9 --zi 255`) in my projects, but executing oxipng in the project root also includes screenshots and other large images. I'd love to be able to exclude larger images from those extreme runs, as they take forever.

### Details 

I suggest adding `--max-input` or something similar, and add different types for limits:
- `--max-input = 1000KB`, we do filesize
- `--max-input = 1000*500` or `--max-input = 1000*500px`, we check x dimension (1000px), and y dimension (500px)
- `--max-input = 1000`, we assume 1000px for both dimesions, so same as `--max-input = 1000x1000px`
- `--max-input = 1000^2`, we use a max pixel total of 1000000px. So anything with less pixels matches - it would match images with max 500\*2000px, 2000\*500px, 1000\*1000px, 10\*100000px and anything smaller

**Comment by andrews05** [comment]
Just a thought of my own: If the concern is performance on larger files, what if the value given is the uncompressed byte size? This would essentially equal width * height * depth. It would give more consistent results over images of different colour types/depths if the dimensions aren't actually of interest here.

**Comment by Maingron** [comment]
My knowledge about PNGs doesn't go this deep, but I assume it'd be similar to a BMP?

**Generally I think this is a good thought**, but would the average user understand the concept whereas they're used to default units like dimensions or filesize. Is there a way to express uncompressed raw size in a simple way?

Also what about 8 bit PNG vs 16 bit PNG vs 24 bit vs 32 bit PNG? How will you compensate for missing channels, especially alpha channel? 
Pre-Oxipng PNG might have an alpha channel with only `0xFF`s, so 32 bit
After-Oxipng PNG might have the alpha channel stripped

**Comment by andrews05** [comment]
Yeah, that is a good point. Hypothetically the check could be done *after* reductions had been performed, but this would be notably more complex. The advantage of doing it as soon as the header has been read is that we wouldn't even need to *de*compress the image data, so it would be effective at limiting memory usage as well (providing some protection from "png bombs" mentioned in the initial post).

I'm sure we could work out how to explain it or provide guidance to users.

**Comment by Maingron** [comment]
I assume you already get dimensions before starting processing anyway -

Would it be difficult to implement multiple units, like px\*px, filesize, and then your idea too, maybe as 10000bpppx+a (bit per pixel \* pixels +**a**lpha flag) or something?
I think all of these have good use cases. While px*px and filesize are the more traditional, well understood and simple ones, the bpppx (todo: find better unit name) brings more power.

I assume by `the check could be done after reductions had been performed` you mean an estimate of how big it would be after reduction, which is a very quick operation?

**Comment by TPS** [comment]
>Hypothetically the check could be done *after* reductions had been performed, but this would be notably more complex.


> I assume by `the check could be done after reductions had been performed` you mean an estimate of how big it would be after reduction, which is a very quick operation?


I understand these POVs to have a completely different comprehension of the costs involved.  Could we have this point clarified?

**Comment by AlexTMjugador** [comment]
Even though #732 was merged, I'm thinking that OxiPNG could technically do more by e.g. refusing to deal with PNG files that have a compression bomb in their compressed text chunks when copying them, as that could cause downstream applications to choke at the generated files. But overall I see it as an effective, low-effort start to having OxiPNG RAM and CPU usage somewhat bounded for most files. If anyone is interested in discussing more resource limiting avenues, please feel free to continue using this thread.

**Comment by andrews05** [comment]
True, we can always reopen if there are still valid concerns. Of course, stripping is an effective way to deal with malicious text chunks.

@Maingron Just to follow up with your last comment, I'd rather not complicate things by implementing multiple ways to tackle the same problem. The method that's been implemented in #732 is the most straightforward solution. Dealing with dimensions can cause both complexity and confusion due to there being two values.

**Comment by Maingron** [comment]
I totally get the complexity part!
I do think however that the max-filesize thing technically goes against Oxipng's idea. After all you're trying to shrink files that are too big, not leaving them as big as they are.
It's definitely MUCH better than nothing, or even better than something really complex, but I think having the ability to also input a dimensional limit is as important as being able to input max-filesize.

If you're still up for it, I suggest either adding `--max-dimensions="100x100"` or adding the same functionality to `--max-size`, but leaving this semi-undocumented.

### Thinking about it by definition, 
**dimensions is different to size**, so splitting the flags into (file-)size and (image-)dimensions would totally make sense to me! 

It'd probably be fine to just add "100x100" and to leave out `"100 on the bigger side **times** I don't care"`, since user can just run command twice `"100x99999"` and `"99999x100"`.
**Although** now in this case it would be mandatory to also have a `--min-dimensions` to avoid duplicate work. I do think `--min-dimensions` would **not** justify THIS GitHub issue, since it's a topic by itself and would have its own use cases.

----

I might suggest `--output-path` and `--min-dimensions` in new issues, but I'll spare you creating issues for that. If you do feel like those 2 flags would be nice to implement, let me know and I'll create one or two issues 🫣   
(`--output-path` could also help with duplicate work, hence the mention)

**Comment by ace-dent** [comment]
@Maingron - I also previously wanted image dimensions, since brute force trials are more expensive per row (pixel height) rather than per column (width). However, most users can approximate dimensions with a maximum *uncompressed* size (please check the example*). For very specific uses, you can get image dimensions from `pngcheck`... or even verbose `oxipng -v` will report dimensions that can be piped to your program / script.

IMHO the approach taken by @andrews05 is going to satisfy the majority of users.

\* Example:
 for my script ideally I want to limit graphics to 8x8 images... if I assume 4 bytes per pixel (RGBA), that's 256 bytes.

@andrews05 - what happens if the image is just B/W? Will it be reported as grayscale 1 byte/pixel uncompressed? For that matter, does alpha channel get counted?

**Comment by andrews05** [comment]
@Maingron I think there might be some confusion here.
Firstly, let me clarify the new option `--max-raw-size` by quoting the help:
> Maximum size to allow for the input image. If the raw, decompressed image data (or the \
file size) of the image exceeds this size, it will be skipped. This is useful for limiting \
memory usage or avoiding long processing times on large images. The value may be specified \
with a unit suffix such as k, KB, m, MB, etc.
The decompressed size of an image is roughly equal to width * height * bit-depth / 8. E.g. \
a 1920x1080 image with 24-bit color depth would be roughly 6MB.

Note it's about the size of the image data, not the file size (although it does check that too as an early abort).

Secondly, I'm a little confused by your comment:
> I do think however that the max-filesize thing technically goes against Oxipng's idea. After all you're trying to shrink files that are too big, not leaving them as big as they are.

While earlier you said:
> I'd love to be able to exclude larger images from those extreme runs, as they take forever.

The topic at hand is indeed about skipping certain files if they might take too long or require too much memory to process. Of course, you may prefer to start by using a fast preset over all your images, then following up with a slower one combined with `--max-raw-size`.

I guess I'm not really understanding why the dimensions are actually important. The size of the image data is more relevant to the required memory and processing time than the specific dimensions. I see the dimensions as merely an alternative way of conveying the same basic idea, except it doesn't take bit-depth into account. With that in mind, and given the complexities around it that you've already demonstrated in your own post, I'm afraid the answer is still a no from me.

Lastly, we do already have an `--out` option. No need to create a new issue for that 😉

**Comment by andrews05** [comment]
@ace-dent B/W could be represented in any bit-depth - if your input is RGBA then it will still be 32-bit.
